ouroboros-consensus/test/infra-test/Test/Ouroboros/Consensus/PeerSimulator/Run.hs

```diff
@@ -0,0 +1,324 @@
+      threadDelay 0.100
+      trace $ condense pid ++ "'s tick is now done."
+    Nothing -> error "“The impossible happened,” as GHC would say."
+  where
+    trace = traceUnitWith tracer "Scheduler"
+
+-- | Iterate over a 'PointSchedule', sending each tick to the associated peer in turn,
+-- giving each peer a chunk of computation time, sequentially, until it satisfies the
+-- conditions given by the tick.
+-- This usually means for the ChainSync server to have sent the target header to the
+-- client.
+runScheduler ::
+  IOLike m =>
+  Tracer m String ->
+  PointSchedule ->
+  Map PeerId (PeerResources m) ->
+  m ()
+runScheduler tracer (PointSchedule ps _) peers = do
+  traceWith tracer "Schedule is:"
+  for_ ps  $ \tick -> traceWith tracer $ "  " ++ condense tick
+  traceWith tracer "--------------------------------------------------------------------------------"
+  traceWith tracer "» Time says “Let there be”"
+  traceWith tracer "» every moment and instantly"
+  traceWith tracer "» there is space and the radiance"
+  traceWith tracer "» of each bright galaxy."
+  traceWith tracer "--------------------------------------------------------------------------------"
+  for_ ps (dispatchTick tracer peers)
+  traceWith tracer "--------------------------------------------------------------------------------"
+  traceWith tracer "» A Clock stopped -"
+  traceWith tracer "» Not the Mantel's -"
+  traceWith tracer "» Geneva's farthest skill"
+  traceWith tracer "» Can't put the puppet bowing"
+  traceWith tracer "» That just now dangled still -"
+
+-- | Construct STM resources, set up ChainSync and BlockFetch threads, and
+-- send all ticks in a 'PointSchedule' to all given peers in turn.
+--
+-- REVIEW: We could extract the PeerIds from the point schedule.
+-- As it is, we could run protocols for only a subset of peers, but that
+-- would cause dispatchTick to crash.
+-- Is this a potential use case?
+runPointSchedule ::
+  (IOLike m, MonadTime m, MonadTimer m) =>
+  SecurityParam ->
+  Asc ->
+  PointSchedule ->
+  Tracer m String ->
+  BlockTree TestBlock ->
+  [PeerId] ->
+  m (Either (NonEmpty SomeException) TestFragH)
+runPointSchedule k asc pointSchedule tracer blockTree peers =
+  withRegistry $ \registry -> do
+    resources <- makePeersResources tracer blockTree peers
+    let candidates = srCandidateFragment . prShared <$> resources
+    traceWith tracer $ "Security param k = " ++ show k
+    chainDb <- mkChainDb tracer candidates config registry
+    fetchClientRegistry <- newFetchClientRegistry
+    let chainDbView = defaultChainDbView chainDb
+    chainSyncThreads <- for resources $ \PeerResources {..} -> do
+      thread <- startChainSyncConnectionThread tracer config asc chainDbView fetchClientRegistry prShared prChainSync
+      PeerSimulator.BlockFetch.startKeepAliveThread registry fetchClientRegistry (srPeerId prShared)
+      pure thread
+    for_ resources $ \PeerResources {..} ->
```

Comment:
Not something required to merge, but I dislike the implicit variables brought into scope. I find easier to learn where `prShared` comes from if the pattern is written like:
```suggestion
    for_ resources $ \PeerResources {prShared} ->
```

Reply:
I agree with this; more explicit more better!